src/java.base/share/classes/java/lang/StringFormat.java

```diff
@@ -0,0 +1,327 @@
+            String s = null;
+            if (args.length == 1) {
+                s = format1(format, args[0]);
+            } else if (args.length == 2) {
+                s = format2(format, args[0], args[1]);
+            }
+            if (s != null) {
+                return s;
+            }
+        }
+
+        return new Formatter().format(format, args).toString();
+    }
+
+    private static String format1(String format, Object arg) {
+        int max = format.length();
+        int off = format.indexOf('%');
+        if (off == -1 || off + 1 == max) {
+            return null;
+        }
+
+        if (off + 2 < max) {
+            if (format.indexOf('%', off + 2) != -1) {
+                return null;
+            }
+        }
+
+        char conv = format.charAt(off + 1);
+        int width = 0;
+        if (conv >= '1' && conv <= '9') {
+            width = conv - '0';
+            if (off + 2 < max) {
+                conv = format.charAt(off + 2);
+            }
+        }
+
+        if (conv == STRING && isBigInt(arg)) {
+            conv = DECIMAL_INTEGER;
+        }
+
+        if (conv == STRING) {
+            arg = String.valueOf(arg);
+        }
+
+        int size = stringSize(conv, arg);
+        if (size == -1) {
+            return null;
+        }
+        return format1(format, off, conv, arg, width, size);
+    }
+
+    private static String format2(String format, Object arg0, Object arg1) {
+        int max = format.length();
+        int off0 = format.indexOf('%');
+        if (off0 == -1 || off0 + 1 == max) {
+            return null;
+        }
+        char conv0 = format.charAt(off0 + 1);
+        int width0 = 0;
+        if (conv0 >= '1' && conv0 <= '9') {
+            width0 = conv0 - '0';
+            if (off0 + 2 < max) {
+                conv0 = format.charAt(off0 + 2);
+            }
+        }
+
+        int off1 = format.indexOf('%', off0 + 1);
+        if (off1 == -1 || off1 + 1 == max) {
+            return null;
+        }
+        char conv1 = format.charAt(off1 + 1);
+        int width1 = 0;
+        if (conv1 >= '1' && conv1 <= '9') {
+            width1 = conv1 - '0';
+            if (off1 + 2 < max) {
+                conv1 = format.charAt(off1 + 2);
+            }
+        }
+
+        if (off1 + 2 < max) {
+            if (format.indexOf('%', off1 + 2) != -1) {
+                return null;
+            }
+        }
+
+        byte coder = format.coder();
+        if (conv0 == STRING && isBigInt(arg0)) {
+            conv0 = DECIMAL_INTEGER;
+        }
+        if (conv1 == STRING && isBigInt(arg1)) {
+            conv1 = DECIMAL_INTEGER;
+        }
+
+        String str;
+        if (conv0 == STRING) {
+            str = String.valueOf(arg0);
+            coder |= str.coder();
+            arg0 = str;
+        }
+        if (conv1 == STRING) {
+            str = String.valueOf(arg1);
+            coder |= str.coder();
+            arg1 = String.valueOf(str);
+        }
```

Comment:
similar to above, should these conditions be consolidated?

```suggestion
        String str;
        byte coder = format.coder();
        if (conv0 == STRING) {
            if (isBigInt(arg0)) {
                conv0 = DECIMAL_INTEGER;
            } else {
                str = String.valueOf(arg0);
                coder |= str.coder();
                arg0 = str;
            }
        }

        if (conv1 == STRING) {
            if (isBigInt(arg1)) {
                conv1 = DECIMAL_INTEGER;
            } else {
                str = String.valueOf(arg1);
                coder |= str.coder();
                arg1 = String.valueOf(str);
            }
        }
```